pkg/cmd/cli/cmd/newapp.go

```diff
@@ -271,15 +274,56 @@ func setLabels(labels map[string]string, result *newcmd.AppResult) error {
 	return nil
 }
 
+// isInvalidTriggerError returns true if the given error is
+// a validation error that contains 'trigger' in its error message.
+// This error is returned from an older server when
+// it doesn't recognize a newer trigger type.
+func isInvalidTriggerError(err error) bool {
+	if !kapierrors.IsInvalid(err) {
+		return false
+	}
+	statusErr, ok := err.(*kapierrors.StatusError)
+	if !ok {
+		return false
+	}
+	return strings.Contains(statusErr.Status().Message, "trigger")
+}
+
+// retryBuildConfig determines if the given error is caused by an invalid trigger
+// error on a BuildConfig. If that is the case, it will remove all triggers with a
+// type that is not in the whitelist for an older server.
+func retryBuildConfig(info *resource.Info, err error) runtime.Object {
+	triggerTypeWhiteList := map[buildapi.BuildTriggerType]struct{}{
+		buildapi.GitHubWebHookBuildTriggerType:  struct{}{},
+		buildapi.GenericWebHookBuildTriggerType: struct{}{},
+		buildapi.ImageChangeBuildTriggerType:    struct{}{},
```

Comment:
I figured out I introduced the validation error on unknown types in #2750 :disappointed:
We fixed it in #4556.

@liggitt this `retryBuildConfig` is something very specific to workaround the situation when you have a new template and new client talking to an old server that throws validation error on unknown trigger types.
We know that a server that throws that error must only know about 3 trigger types: GitHubWebhook, GenericWebhook and ImageChange, therefore we retry to create the BuildConfig filtering out any trigger other than those 3 types.

On (de)serialization level, Go's json parser will ignore fields that are not in the struct we deserialiaze into.

Given a JSON like:
```json
{
  "triggers": [
    {
      "type": "Foo",
      "fooTrigger": {}
    }
  ]
}
```

And if we don't have a `fooTrigger` field, we'll end up with:

```go
[]Trigger{
  {Type: "Foo"},
}
```

I think that's fine, our code ignores that trigger and there's no way to have nil pointer access on that, since we don't know about Foo and we'll never try to read `fooTrigger`.
In case we add a `fooTrigger` field later, and an old template is wrong, or you have an old client that cannot deserialize that trigger properly, it will fail validation for that specific trigger.

The latter might puzzle the user... she looks at her template and see a perfectly crafted type Foo trigger and fooTrigger parameters, yet `oc new-app` throws an error saying she's got an invalid trigger (because the client ignored the `fooTrigger` part it doesn't know about). How do we help her figure out that she needs a new version of the `oc` client?